File: Fleet/CoreExtensions/UIAlertAction+Fleet.m
#import "UIAlertAction+Fleet.h"
#import <objc/runtime.h>

@interface ObjectifiedBlock : NSObject

@property (nonatomic, copy) void (^block)(UIAlertAction *);

- (instancetype)initWithBlock:(void (^)(UIAlertAction *))block;

@end

@implementation ObjectifiedBlock

- (instancetype)initWithBlock:(void (^)(UIAlertAction *))block {
    if (self == [super init]) {
        self.block = block;
    }

    return self;
}

@end

BOOL didSwizzleUIAlertAction = NO;
unsigned int handlerAssociatedKey = 0;

@interface UIAlertAction (FleetPrivate)

@property (nonatomic, copy) void (^fleet_property_handler)(UIAlertAction *);

@end

@implementation UIAlertAction (FleetPrivate)

+ (void)initialize {
    if (self == [UIAlertAction class]) {
        if (!didSwizzleUIAlertAction) {
            [self swizzleHandlerSettler];
            didSwizzleUIAlertAction = YES;
        }
    }
}

+ (void)swizzleHandlerSettler {
    SEL originalSelector = NSSelectorFromString(@"setHandler:");
    SEL swizzledSelector = @selector(fleet_setHandler:);

    Method originalMethod = class_getInstanceMethod(self, originalSelector);
    Method swizzledMethod = class_getInstanceMethod(self, swizzledSelector);

    method_exchangeImplementations(originalMethod, swizzledMethod);
}

- (void)fleet_setHandler:(void (^)(UIAlertAction *))handler {
    self.fleet_property_handler = handler;
    [self fleet_setHandler: handler];
}

- (void (^)(UIAlertAction *))fleet_property_handler {
    ObjectifiedBlock *block = objc_getAssociatedObject(self, &handlerAssociatedKey);
    return [block block];
}

- (void)setFleet_property_handler:(void (^)(UIAlertAction *))fleet_property_handler {
    ObjectifiedBlock *block = [[ObjectifiedBlock alloc] initWithBlock: fleet_property_handler];
    objc_setAssociatedObject(self, &handlerAssociatedKey, block, OBJC_ASSOCIATION_RETAIN);
}

@end

File: Fleet/CoreExtensions/UIAlertAction+Fleet.swift
import UIKit
import ObjectiveC

private var handlerAssociatedKey: UInt = 0

fileprivate var didSwizzle = false

@objc private class ObjectifiedBlock: NSObject {
    var block: ((UIAlertAction) -> Void)?

    init(block: ((UIAlertAction) -> Void)?) {
        self.block = block
    }
}

extension UIAlertAction {
    var handler: ((UIAlertAction) -> Void)? {
        get {
            return fleet_property_handler
        }
    }

    override open class func initialize() {
        super.initialize()
        if !didSwizzle {
            swizzleHandlerSetter()
            didSwizzle = true
        }
    }

    fileprivate class func swizzleHandlerSetter() {
        let originalSelector = Selector(("setHandler:"))
        let swizzledSelector = #selector(UIAlertAction.fleet_setHandler(_:))

        let originalMethod = class_getInstanceMethod(self, originalSelector)
        let swizzledMethod = class_getInstanceMethod(self, swizzledSelector)

        method_exchangeImplementations(originalMethod, swizzledMethod)
    }

    func fleet_setHandler(_ handler: ((UIAlertAction) -> Void)?) {
        fleet_property_handler = handler
        fleet_setHandler(handler)
    }

    fileprivate var fleet_property_handler: ((UIAlertAction) -> Void)? {
        get {
            let block = objc_getAssociatedObject(self, &handlerAssociatedKey) as? ObjectifiedBlock
            return block?.block
        }

        set {
            let block = ObjectifiedBlock(block: newValue)
            objc_setAssociatedObject(self, &handlerAssociatedKey, block, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN)
        }
    }
}
